### PR TITLE
[REFACTOR][Triton] Per-token-head: first-chunk fast-path, mixed batch split, fused K+V, dedicated decode kernel

### PR DESCRIPTION
**Pull request by JartX**
<html><head></head><body><h1>Fast paths and fused cache layout for <code>per_token_head</code> KV quantization</h1>
<h2>Background</h2>
<p>When <code>int8_per_token_head</code> or <code>fp8_per_token_head</code> quantization is active, <code>unified_attention</code> becomes the hot path for both prefill and decode. It reads the quantized cache, loads per-<code>(block, slot, head)</code> scales inside the tile loop, and dequantizes K/V on the fly. That overhead is worth paying when attending to previously cached context — but it dominates two cases where a cheaper path exists:</p>
<ul>
<li><strong>First-chunk prefill</strong>: the raw Q/K/V tensors are already in registers from the linear projection, and the cache doesn't hold anything useful yet.</li>
<li><strong>Decode / small continuation prefill</strong>: <code>unified_attention</code> is a generic kernel that carries dead branches for alibi, sinks, softcap, sliding window, and mm_prefix, and has to coexist with non-quantized and fp8-per-tensor paths — all of which hurts occupancy.</li>
</ul>
<p>There's also a layout problem: the <code>per_token_head</code> cache places K and V in separate halves of the leading dimension, so reading K and V for the same position touches two distant memory regions (~100 KB apart) and wastes cacheline bandwidth.</p>
<hr>
<h2>What this PR does</h2>
<p>Three cooperating optimizations, all gated to <code>per_token_head</code> — zero impact on any other KV cache dtype.</p>
<h3>1. Fused K+V cache layout (<code>per_token_head</code> only)</h3>
<p><code>get_kv_cache_shape</code> now returns <code>(num_blocks, 1, block_size, num_kv_heads, 2 * (head_size + scale_pad))</code>. The size-1 dummy dim preserves the 5D rank the allocator expects, while the inner dim packs K and V contiguously as <code>[K_data | K_scale | V_data | V_scale]</code>. K and V for the same <code>(block, pos, head)</code> end up bytes apart instead of ~100 KB apart, which immediately improves L2 hit rate when <code>unified_attention</code> reads them back-to-back.</p>
<p><code>_ensure_fused_cache_views</code> carves four zero-copy strided views (K, V, K_scale, V_scale) out of the single cache tensor. Existing Triton kernels — both the store kernel and <code>unified_attention</code> — consume them via <code>.stride()</code> with no kernel-side changes.</p>
<h3>2. Fast-paths in <code>TritonAttentionImpl.forward</code></h3>
<p>Three dispatch cases short-circuit the generic <code>unified_attention</code>:</p>

Case | Condition | Action
-- | -- | --
1 | all_pure_first_prefill == True, no decodes | Run context_attention_fwd on raw Q/K/V. No cache reads.
2 | Mixed first-chunk prefill + decodes | Decodes → unified_attention (paged cache); prefills → context_attention_fwd on raw Q/K/V. Requires reorder_batch_threshold=1, enabled for per_token_head only.
3 | Decode + small continuation (max_query_len ≤ 128) | Dispatch to the dedicated split-KV kernel (see below).

<p><strong>Case 2</strong> additionally requires <code>num_decodes &gt; 0</code> and the prefill slice being entirely first-chunk. <strong>Case 3</strong> additionally requires <code>max_query_len ≤ 128</code> and <code>sliding_window == (-1, -1)</code>.</p>
<p><code>build_for_cudagraph_capture</code> resets <code>all_pure_first_prefill</code> and <code>prefill_is_first_chunk</code> to <code>False</code> so the prefill fast-paths are never baked into a captured graph.</p>
<h2>Numbers</h2>
<p>Initial measurement (Case 1 only) showed ~40% of <code>per_token_head</code> calls taking the fast-path under agent-style traffic:</p>
<pre><code>fast=370  slow=400  ratio=0.481
fast=400  slow=500  ratio=0.444
fast=420  slow=600  ratio=0.412
</code></pre>
<p>Case 2 additionally catches mixed batches (new user prefill while others decode), and Case 3 + the fused layout accelerate the remaining decode and continuation traffic. The combined end-to-end speedup is most visible on prefill-heavy workloads with long contexts and on steady-state decode throughput.</p></body></html>

### CODE REVIEW

File: vllm/v1/attention/backends/triton_attn.py [hunk 1]
```diff
@@ -557,6 +576,31 @@ def forward(
...
+        if (
+            self._is_per_token_head_quant
+            and attn_metadata.all_pure_first_prefill
+            and self.alibi_slopes is None
+            and not self.use_alibi_sqrt
+            and self.sinks is None
+            and self.logits_soft_cap == 0
+            and attn_metadata.mm_prefix_range_tensor is None
+            and output_scale is None
+        ):
```
[the remark this answers is not in the document]
Comment (JartX) [reply vllm/v1/attention/backends/triton_attn.py:777]: Thanks! The dtype coercion is applied in the latest commit.

Regarding expanding the fast-path to non-quantized dtypes: for FP16/BF16 the unified_attention path already avoids the per-tile scale loads and dequantization that motivate this PR, so the expected win is much smaller (possibly zero or negative, since unified_attention has better tiling for long prefills). Paged-vs-contiguous is a second-order concern in the first-chunk prefill case because the tokens come directly from the linear projection regardless. I'd rather keep this PR scoped to the quantized path where the win is measured and clear; happy to explore the general case in a follow-up

**Comment by JartX** [comment]
@huangzhilin-hzl 

Would you be so kind as to run the same tests as on the TurboQuant test, both from the main test and from this PR test on the high-end graphics card? Thank you very much!

It should be most noticeable in long contexts
